src/item/ArmorTrimRegistry.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\item;
+
+use pocketmine\data\bedrock\item\ItemTypeNames as Names;
+use pocketmine\item\ItemTypeIds as Ids;
+use pocketmine\utils\SingletonTrait;
+use pocketmine\utils\TextFormat;
+
+final class ArmorTrimRegistry{
+	use SingletonTrait;
+
+	/**
+	 * @var ArmorTrimMaterial[] $materials
+	 * @phpstan-var array<string, ArmorTrimMaterial>
+	 */
+	private array $materials = [];
+
+	/**
+	 * @var ArmorTrimMaterial[] $patterns,
+	 * @phpstan-var array<int, ArmorTrimMaterial>
+	 */
+	private array $idToMaterialMap = [];
+
+	/**
+	 * @var ArmorTrimPattern[] $patterns
+	 * @phpstan-var array<string, ArmorTrimPattern>
+	 */
+	private array $patterns = [];
+
+	/**
+	 * @var ArmorTrimPattern[] $patterns,
+	 * @phpstan-var array<int, ArmorTrimPattern>
+	 */
+	private array $idToPatternMap = [];
+
+	public function __construct(){
+		$this->registerDefaultMaterials();
+		$this->registerDefaultPatterns();
+	}
+
+	private function registerDefaultMaterials() : void{
+		$this->registerMaterial(new ArmorTrimMaterial(ArmorTrimMaterial::AMETHYST, TextFormat::MATERIAL_AMETHYST, Names::AMETHYST_SHARD, Ids::AMETHYST_SHARD));
```

Comment:
As I said in discord, it would be better to pass an Item and then serialise it when sending TrimData instead of passing type id and serialised item name

Reply:
Agreed.

I also don't like that this class is the only way to get an `ArmorTrimMaterial` instance or `ArmorTrimPattern`. Feels very `ItemFactory`-esque in that it forces the use of ID constants and obnoxiously long call syntax that encourages plugin devs to hardcode stuff.

`ArmorTrimRegistry::getInstance()->getMaterial(ArmorTrimMaterial::COPPER)`
could be more like
`VanillaArmorTrimMaterials::COPPER()`

The job of classes like this should be to load & save data from storage, not to be used as APIs. It should be in the `pocketmine\data\item` package.